common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -225,76 +246,137 @@ public String encrypt(final String clearText)
+            return EncryptionType.UNENCRYPTED;
+        }
 
         try {
-            return decryptWithPackageName(bytes, packageName);
-        } catch (GeneralSecurityException | IOException e) {
-            Logger.error(
-                    TAG + methodName,
-                    "Failed to decrypt with package name: "
-                            + packageName,
-                    e
+            final String keyVersion = new String(
+                    bytes,
+                    0,
+                    KEY_VERSION_BLOB_LENGTH,
+                    AuthenticationConstants.ENCODING_UTF8
             );
 
-            if (!BROKER_PACKAGE_NAMES.contains(packageName)) {
-                // If we are not the broker, do not retry, immediately throw.
-                throw e;
+            if (VERSION_USER_DEFINED.equalsIgnoreCase(keyVersion)) {
+                return EncryptionType.USER_DEFINED;
+            } else if (VERSION_ANDROID_KEY_STORE.equalsIgnoreCase(keyVersion)) {
+                return EncryptionType.ANDROID_KEY_STORE;
             }
+        } catch (UnsupportedEncodingException e) {
+            Logger.error(TAG + methodName, "Failed to extract keyVersion.", e);
+            throw e;
+        }
 
-            // We are the broker... try with the other broker's package name
-            final String nextPackageToUse =
-                    packageName.equals(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)
-                            ? COMPANY_PORTAL_APP_PACKAGE_NAME
-                            : AZURE_AUTHENTICATOR_APP_PACKAGE_NAME;
+        return EncryptionType.UNENCRYPTED;
+    }
 
-            Logger.warn(
-                    TAG + methodName,
-                    "We are the broker. Retrying with package name: "
-                            + nextPackageToUse
-            );
+    private byte[] getByteArrayFromEncryptedBlob(@NonNull final String encryptedBlob) {
+        int encodeVersionLength = encryptedBlob.charAt(0) - 'a';
+        validateEncodeVersion(encryptedBlob, encodeVersionLength);
+
+        return Base64.decode(
+                encryptedBlob.substring(1 + encodeVersionLength),
+                Base64.DEFAULT
+        );
+    }
 
-            return decryptWithPackageName(bytes, nextPackageToUse);
+    /**
+     * Get all the key type that could be potential candidates for decryption.
+     **/
+    @NonNull
+    private List<KeyType> getKeysForEncryptionType(@NonNull final String encryptedBlob,
+                                                   @NonNull final String packageName) throws IOException {
+        final String methodName = ":initializeDecryptionKeyTypeList";
+        List<KeyType> keyTypeList = new ArrayList<>();
+
+        EncryptionType encryptionType = getEncryptionType(encryptedBlob);
+
+        if (encryptionType == EncryptionType.USER_DEFINED) {
+            if (COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+                keyTypeList.add(KeyType.LEGACY_COMPANY_PORTAL_KEY);
```

Comment:
Should these add try keystore too?

Reply:
I think we're good. In this case we know that the key is not encrypted by the Keystore-encrypted symmetric key.